### ISSUE TITLE
[processingjs.org] param() incorrectly marked as not implemented

### ISSUE THREAD

**Issue by matteosistisette** [processing-js/processing-js#140]
I'm reporting this here because https://github.com/processing-js/processing-js.github.io doesn't seem to have an issue tracker. 

At http://processingjs.org/reference/ param() is marked as "not yet implemented" (and consistently, its documentation page is missing).
But the method is implemented and it works.

I hope you're not planning to remove it.

**Comment by Pomax** [comment]
Hm, this might have simply slipped through the cracks, it's a function I've not seen use in (literally) years. That said, Processing itself no longer supports it, so in order to stay on parity with Processing, we're going to *have* to remove it at some point. 

What are you using it for?

**Comment by GoToLoop** [comment]
I don't see the need of removing any working feature!
Transpiling Java Mode(P5)  to JS Mode (PJS) demands that JS Mode side have all Java Mode API.
But JS Mode can perfectly have more features of their own!

**Comment by matteosistisette** [comment]
@Pomax Processing no longer supports param() only because Processing no longer supports exporting Java applets, which is to say, Processing no longer supports any way of embedding a sketch into a web page, other than using JavaScript mode (which is sad, but the Java plugin for browsers is going to die anyway).

Processing has stopped supporting exporting java applets, and hence publishing sketches into html pages, precisely because ProcessingJS has provided a seamless alternative to do that. Now if you remove it from Processing JS, you're going to cripple Processing/ProcessingJS when it comes to publishing a sketch online, as using parameters is a vital feature (see below).

Actually, Processing has made a mistake in not providing a dummy implementation for param(), to be used for testing, when they dropped support for exporting applets, because they this prevents us from even compiling in java mode a sketch that uses this feature, which is the only way to receive a parameter from the page where the sketch is going to be exported for web (and the only way to export it is ProcessingJS).

I don't understand what you mean when you say "it's a function I've not seen use in (literally) years". How do you "watch" what features are and aren't used? Does processing.js have some hidden piece of code that secretly sends you a copy of every sketch that is ever run, and you actually review all of them? Just kidding.


* USE CASES*
What do I use it for? Isn't it obvious? (by the way, just about EVERY kind of plugin - Flash, Unity, Java...- that allows to embed objects that run code in an html page provide a way of passing them parameters)

The most obvious example:
a) multiple instances of the same sketch in a single page, each with different parameters.
such as http://matteosistisette.com/upf-redes-x/ejemplos/BDD/processing_lienzos_usuarios/
(this is an example for teaching so yes, it's very simple and naive)
And another use case:
b) a sketch hosted in a .php page where some parameter for the sketch is determined from PHP when the page is loaded (obviously replace PHP with any other server-side language of your choice)

Yes, there are workarounds for accomplishing the same, most of which would be unnecessarily convoluted, cumbersome and  unelegant  hacks, especially for case (a).


About this:
> Processing itself no longer supports it, so in order to stay on parity with Processing, we're going 
> to have to remove it at some point.

As GoToLoop pointed out that makes no sense.
I appreciate your concern for "parity" if that means that everything that works in Processing needs to work in ProcessingJS, but if you're going to enforce the other way round (from the PJS side alone, i.e. ensuring that anything that doesn't work in Processing won't work in PJS either even when it could and did), then both Processing and Processing JS are going to die. And I'm serious.

WARNING: drifting OT from now on

Processing without PJS has condemned itself to be offline-only (granted, it can send and receive data to/from the Internet, but a sketch cannot be hosted online!!) the moment they dropped support for exporting java applets. That, together with other things such as lack of support for asynchronous I/O, was confining Processing into the XX century (actually, to pre-1990s), if it wasn't thanks to PJS which had/has the opportunity to bring Processing back to 21st century. 
But if you keep this both-ways-100%-parity approach, then Processing will quickly become obsolete because 
- Processing (non JS) will become a useless platform for anything that needs to be published online, which is nowadays pretty much anything
- Processing will likely remain stuck to its Java legacy, making its API obsolete in some areas (such as synchronous I/O), even areas where improvements are perfectly possible (but not as easy as in JS)
- ProcessingJS won't fill those gaps either, if you keep it stuck to parity with Processing

There are improvements that are very easy to achieve in JS (actually, there are things where PJS is doing a great active effort in crippling itself in order to remain compatible with P5, and I'm not saying this as criticism, as I understand the importance of compatibility).

So, the only way I can foresee a future in which Processing won't die to be superseded by other amazing JS-centered libraries and frameworks, is this:
- PJS shoudl and could introduce new features (without breaking compatibility with P5's existing API) which can easily be implemented in JS, leveraging the power of JS and bringing PJS to the next step, keeping it current with the present
- P5 would then be stimulated to follow that, and implement those same improvements back into Processing (which in many case is possible, though not as easy as in JS)

In other words, you don't always have to wait for P5 to do the first step.

Processing is a great language for learning programming (which is what it was designed for in the first place), but in the current world, dominated by html5-javascript and where being offline is no longer an option, Processing will eventually become great-for-learning-but-useless-to-do-real-things.
PJS is the only opportunity for Processing to avoid that.

(as a side note, I find some irony in the contraddiction between your obsession with parity and your lack of interest in fixing the issue with loadXML we recently talked about, which actually breaks parity, for the sake of BC with some old broken no-longer-supported versions of the API)

**Comment by Pomax** [comment]
whoa, back up here, both of you. There is a huge difference between "removing an obsolete function" and "removing the ability to do [...]". Processing.js runs in a JavaScript environment, and it is trivial to load data into a sketch (or get it out of a sketch) using plain old JS.

Like with all programming APIs, update your methodology once parts of APIs go obsolete: we're still in JavaScript context, and still in a browser, so getting data into and out of a sketch is super simple. If you write code that can only run in JS context already anyway, then make use of JS to do data loading for you, by using things like element attribute extraction, reading in .innerHTML values, and using XHR data retrieval for data from URLs. 

- Will param() always be around? Probably not.
- Should you stop using it? Absolutely: there are better ways in modern JS to do what it was used for in Java applets.
- Is it going to disappear really soon? Unlikely, it's something that'll likely get removed when Pjs is next overhauled, and there is no timeline for that in the slightest.

Ditching the habit of use of an obsolete function isn't the end of the world, especially when what it does can already be done better by taking advantage of the runtime environment.

**Comment by matteosistisette** [comment]
There's nothing obsolete about param().
It's doing exactly what you suggest us to do. But it is already there, nicely wrapped into an easy-to-use function.

"we're still in JavaScript context, and still in a browser, so getting data into and out of a sketch is super simple. If you write code that can only run in JS context already anyway, then make use of JS to do data loading for you"
And what would be one the nicest and most elegant ways to do that? Put that into a data-xxx attribute of the canvas, and write a nice wrapper function in our sketch to extract it from that data attribute. And that's exactly what processingjs's param() does. So there's nothing obsolete  about it, unless there's something wrong in its _implementation_ which I haven't inspected

**Comment by Pomax** [comment]
I'm glad you're being just as selective as I am. The param() function is obsolete. It's not in the Processing API, and it's not a PJS-specific directive. The functionality is useful, and utterly undocument, so one question you'll want to ask is how you even know it's there. No one who's come to Processing since it was removed from the API will even know it exists, or what it does. So that's one thing.

As for there being something wrong with the implementation: there is indeed. In addition to the `data-processing-...` codepath it also works the way Processing "intended" it to be used, with `<param>` elements inside the `<canvas>`. This is illegal under HTML5 rules, so if we are going to revisit the function, that part'll have to go, at which point it makes far more sense to rip out the function call for param() entirely, simply grep the `data-processing-...` attributes on initialisation, and note that there will be a `sketch.data` or the like available in your sketch's scope.

Having said that, however, you can just as easily put the values you're hardcode into your HTML into its own blah.pde file instead, and then load that alongside your generic "runs on multiple pages" sketch code, e.g. `<canvas data-processing-files="../p5/framework.pde ../p5/SomeClass.pde ./settings.pde">` and simply have a different setting file for each page that has different values. This was already an option when param() *was* part of Processing, and was a generally far robuster approach. It still is.

**Comment by matteosistisette** [comment]
> Like with all programming APIs, update your methodology once parts of APIs go obsolete:

Please don't think that I hate you or something because of that old loadXML() thing, but I can't help noticing _again_ how ironic this statement is, coming from you, given how you refused to fix a fundamental flaw in loadXML() and the XML() constructor for the sake of guess what? backward compatibility with an obsolete API.

**Comment by Pomax** [comment]
indeed, and that's unfortunate and I wish I had time to fix the XML aspect of Pjs, but I don't. I work on an incredible amount of projects for Mozilla as well as other open source initiatives, and all of those are more important to me. Fixing code takes large chunks of days, and I don't have large chunks of days available, which is why I will happily take and review PRs, but will not be doing any writing of those myself until I magically find free days to actively work on Pjs.

**Comment by matteosistisette** [comment]
> As for there being something wrong with the implementation: there is indeed. 

Good, then that's worth fixing.

> In addition to the data-processing-... codepath it also works the way Processing "intended"  it to be used, with <param> elements inside the <canvas>. This is illegal under HTML5 rules, so if we are going to revisit the function, that part'll have to go, 

Agreed

> at which point it makes far more sense to rip out the function call for param() entirely, 

There I don't follow you any more.

I must admit I hadn't thought about the "settings.pde approach", but I still find the data-attribute approach much cleaner (for one thing, if the data had to be dinamic generated from the server side, it would be pretty cumbersome to implement, using e.g. php-generated pde sources; also that means unnecessary extra requests to the server)

> The param() function is obsolete. It's not in the Processing API, and it's not a PJS-specific directive. The functionality is useful, and utterly undocument, so one question you'll want to ask is how you even know it's there. No one who's come to Processing since it was removed from the API will even know it exists, or what it does.

It's true that I only know it's there because I knew about it back from the days when it did exist in Processing. But I found it because I was looking for such a functionality.
If I didn't know it existed, I would be asking for it.
The param() function with its name and history may be obsolete, but the idea of having a function providing that kind of interface (data-someprefix-xxx attributes and a convinient function for easily retrieving data from them) is not an obsolete concept.


> indeed, and that's unfortunate and I wish I had time to fix the XML aspect of Pjs, but I don't.

Ok that makes perfect sense. Back then I was under the impression that you were defending a _choice_ of having XML work the way it works on technical grounds that were simply wrong.

**Comment by Pomax** [comment]
on the "There I don't follow you any more" point, I'm saying that, as param() is not a known function anywhere (as you point out, you had to hunt for it) it makes far more sense to not even bother making it a function that has to be called, simply making it functionality that *always* runs on sketch load, giving you a Pjs-specific global variable like "params" so you can just do this in your code:
```
<canvas data-processing-sources="..." data-processing-stringbuffer="['a','b','c','d'" ...>
```
paired with 
```
void setup() {
  size(params.get('w',100), params.get('h',100)); // retrieval with fallback value
  stringbuffer = params.stringbuffer
  ...
}
```
or
```
void setup() {
  size(params.w, params.h); // much more fragile
  stringbuffer = params.stringbuffer
  ...
}
```
This, to me, makes far more sense than sticking with a historical function for Java applets that doesn't exist anymore in real Processing, and never had any documentation. Putting the same functionality in a different place so that it "just works" specifically for Pjs only sounds like a far better solution to me,

As for taking a shot at the XML fixing: that'd be amazing, I'd be very happy if that got fixed.

**Comment by matteosistisette** [comment]
> it makes far more sense to not even bother making it a function that has to be called, simply making it functionality that always runs on sketch load, giving you a Pjs-specific global variable like "params" so you can just do this in your code

Ohhhhhhhh ok, I missed that part!
That would be great then!

I thought you were talking about getting rid of the feature altogether. Removing the function and not replacing it with anything, leaving it to the user to do the job on their own.

I don't see (but I may be missing) a specific reason why params.get('w') is better than param('w'), but I have nothing against that, and yes, there's no reason for preserving the param() form, other than BC (to be completely fair: this is not BC with a never-documented feature, it's BC with a feature that at some point was documented in Processing and was working in Processing.js; also to be noted that at some point Processing.js used to be bundled with Processing in its distribution, so it was kind of legitimate to rely on Processing's reference to write code which would then run as PJS... this is only to say that if BC was a concern, this would be BC of the legitimate kind, not the dont-break-anything-that-used-to-work-even-if-by-accident kind of BC - but that said, I'm not against breaking BC when it's worth it, especially when we're talking about Processing where every minor release has always broken BC in many ways and for much more dramatic things)

**Comment by GoToLoop** [comment]
* Even though undocumented, **param("")** still exists in Processing v2.2.1!
* And even though export to web applet was officially removed, a tool was created to do that.
* Moreover, http://OpenProcessing.org is still accepting Processing Java Applets as legacy.

**Comment by Pomax** [comment]
There are numerous functions that "exist" in Processing 2 due to how the code has been structured - there's a great number of internal functions that you can actually call because they had to be visible to real API functions, so going on whether you can call it or not is not indicative of whether it's part of the proper Processing API (the general rule is if it's not in the Processing API, it's not technically supported, can't be expected it to work outside of the PDE, and might disappear at any time).

As a former professional Java developer, all I can say about applets still being supported is "that's unfortunate, I hope they turn that off soon". Embedded applets are a bad idea for many reasons, and with the work that's been put into web APIs over the last four years, it's time to take Java out of the browser again.

**Comment by GoToLoop** [comment]
Supported or not, PJS project should strive to get features rather than simply mirror P5 API!
I don't get why PJS can't have more features than P5!  =(

**Comment by Pomax** [comment]
Because that's *the whole point* of Processing.js =)

It's not a general purpose graphics library, it's an implementation of, very specifically, only the core Processing API, and the Processing programming model (which is itself rather different from Java), so that Processing code can run in the browser. If you need to do more than Processing.js offers in your sketches, by all means: shim it using plain old JS, because you can. No one said you had to stick with only Processing.js for what you can make the browser draw, and if you need more, "just add whatever you need"; write some JS to do more than base Pjs can do, and then load that alongside it - much like you would for any other JavaScript library.

To make this as clear as I can, the driving core principle of *any* development on Processing.js itself has always been, and still is, that it's supposed to make your Processing code work on the web (within the limits of technology, e.g. no import parsing because JS can't do anything with compiled java byte code). It was never meant to be an independent library that has a different API from, or richer API than Processing, and in fact any attempt to take it in that direction has been, and will continue to be, a reason to decline patches that try to do so.

The *core* Processing.js library exists solely to make *core* Processing code run in the browser. No more, no less.

(The only "extra bits" that make sense, given that guideline, is things that are necessary to overcome technological limitations, such as image preloading to overcome the difference in synchronous java vs. asynchronous browser behaviour)

**Comment by matteosistisette** [comment]
You've made the intent of Processing.js quite clear and I totally respect that, though it makes me sad.

I'm hoping that someone will work the project and push it in the direction of extending it beyond the limitations of core Processing. Which in turn might encourage people to "backport" some of the improvements to Processing.

Because many of core Processing's limitations come from (a) its Java legacy, which imposes limitations that today are just unbearable, and (b) unavailability of manhours (and perhaps a lack of interest, sadly) from the Processing developers to overcome those limitations; which in Java require more work than they would in JavaScript.

Given how Processing is evolving (or involving) today (more regressions than fixes or new feature at each release; lack of interest in overcoming the limitations we're talking about here, Android mode being currently broken, etc), Processing is already starting to decline. It may be a long long time before it dies, but you can already start to feel the agony. 

ProcessingJS gave me (and I guess many others) the hope that Processing would revive, but as far as the "whole point" of PJS remains "solely to make core Processing code run in the browser" (which as I said I respect, and is a very noble purpose in itself), that won't happen (not impulsed by PJS, anyway).

In PJS it would be *relatively* easy to implement improvements which, if well designed and defined, would serve as a "proof of concept" for Processing to adopt them.

I understand that this is just not part of your project. I'm sad to hear that you will even reject contributions in that direction, but hey, fortunately this is opensource, so anybody may fork this project.


One funny thing is that today you may reject a contribution that adds some amazing feature (consistent with Processing's philosophy and conventions), because it does not exist in Processing, while perhaps the Processing community is dreaming about that very same feature, and some day they'll develop it on their own, and you'll find yourself adding a feature that could have been available years before.

**Comment by GoToLoop** [comment]
1 more question: Would you at least accept code which transpiles useful Java API which aren't present in Processing API?
For example: **HashSet**. **ArrayDeque**, **Rectangle**, etc.?

**Comment by Pomax** [comment]
If it's in the Processing API, then yes. If it's not, there is no reason add it to Processing.js itself, we can just bundle it as an extra.

Maybe this is unknown to either of you, but Processing.js compiles straight to native JavaScript (no, I know, you know that part) but that also means that any Object type you need, you can *literally* define in JS and then use in your sketch.

Need a HashSet?

```
window.HashSet = function() {
};
HashSet.prototype = {
    size: function() { ... },
    ...
}
```
done, you can now use this in your sketch:
```
HashSet set = new HashSet();
...
println(set.size());
...
```
It just works. That's why I don't want to add things to core Processing.js that isn't in Processing: there is *no* reason to. Just write the JS code for it, and ask to have it merged into the extras dir or something, or release it as a thing that can be installed via bower/npm and the entire world can use it? This kind of code doesn't need to be "in" Processing.js to do *exactly* what you want it to.
